### DISCUSSION TITLE
UWB Algorithm

### DISCUSSION THREAD

**Discussion by CalvinKoung** [bitcraze/discussions#1148] [Q&A]
hi, I want to modify the UWB algorithm and I found the uwb_twr_anchor.c this code file and in this code I can saw the uwbTwrAnchorAlgorithm this function. My question is where can I find this uwbTwrAnchorAlgorithm function that I can to read and modify it contents. tks

**Comment by knmcguire** [comment]
uwbTwrAnchorAlgorithm is not a function but a struct but it holds the functions that are important to you. The function connected to .init is [twrAnchorInit](https://github.com/bitcraze/lps-node-firmware/blob/73f8960c728a3030b331b0abe19f73239bf6f38b/src/uwb_twr_anchor.c#L258) and connect to onEvent is [twrAnchorOnEvent](https://github.com/bitcraze/lps-node-firmware/blob/73f8960c728a3030b331b0abe19f73239bf6f38b/src/uwb_twr_anchor.c#L236 ), where the latter reacts on packets received, sent or anything else happening on the ultrawideband module.

In VScode, if you have the c/c++ package installed from microsoft, you can right click on any function or variable and 'go to definition', which will make it easier for you to find the code you are looking for.

**Reply by CalvinKoung** [reply]
so the algorithm is in the this two function ?

**Reply by CalvinKoung** [reply]
And I have one question that if I want to modify the UWB algorithm. which one I need to modify ? the LPS anchor or crazyflie deck computer ? I have some confuse.

**Reply by knmcguire** [reply]
Seems so I think. But just mind that the anchor code is not the only code used for TWR (or TDOA). Technically all the position estimation is done on the Crazyflie itself so you'll also need to look at that code as well. The 'algorithm' is the interplay between the anchors and the tag, if you are talking about the UWB algorithm as a whole.

So I'm assuming that you'd like to change the UWB positioning somehow, or want to get a start at it. If so and you haven't experienced the crazyflie-firmware or any other UWB based positioning before, just a fair warning that it is quite a steep hill to understand the system as a whole This is because you'll need to know 3 things: 1- What the node does with Ultrawideband, 2- what the crazyflie does with UWB, and 3- how state estimation works (for different techniques as well).

Here is a list of documentation to go through to help you to go into the right direction, which I recommend you to through first. Perhaps it will bring you in the right direction. 
- Crazyflie-firmware: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/
- State estimation (which estimates the position) : https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/#extended-kalman-filter
- lps-node-firmwaerhttps://www.bitcraze.io/documentation/repository/lps-node-firmware/master/
- Dev meeting about the LPS system: https://youtu.be/e8TXBsnBuNc

**Reply by CalvinKoung** [reply]
ok, I try my best.

**Reply by CalvinKoung** [reply]
My goal is add the chose anchor this function to filter NLOS signal anchor

**Reply by knmcguire** [reply]
alright good to know! start a new thread if you have any followup question about implementation details on this and we will try to help you point in the right direction.
